### PR TITLE
GPDB GUC dispatching flag refactor

### PR DESCRIPTION
**Pull request by weinan003**
So far, we use to dispatch a guc when creating gang if its flag contains
GUC_GPDB_ADDOPT. It has some drawbacks.
1. This flag left a lots footprint in guc.c, when we diff upstream code.
2. It is easy to forget marking it when new guc introduce.

To deal with these issues, and also this PR is the first step of
[PR7184](https://github.com/greenplum-db/gpdb/pull/7184).

Rename GUC_GPDB_ADDOPT flag as GUC_GPDB_NEED_SYNC, as well as, declare a new
guc flag named GUC_GPDB_NO_SYNC. For the predefine gucs in guc.c and guc_gp.c
must explicitly assign one of these two flags.
For custom guc, give a default flag, GUC_GPDB_NO_SYNC, if it does not
explicit set.

Instead of direct set flag in guc.c, we define two array, sync_guc_names_array
and unsync_guc_names_array.To recode guc names into and automatically
assign flag when the guc object is built in Postgres. Hence, we can
reduce footprint when diffing with upstream guc.c

Idea from Heikki Linnakangas <hlinnakangas@pivotal.io>
Idea from Hubert Zhang <hzhang@pivotal.io>

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [x] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/include/utils/unsync_guc_name.h [hunk 1]
```diff
@@ -0,0 +1,518 @@
...
+		"synchronize_seqscans",
+		"synchronous_commit",
+		"synchronous_standby_names",
+		"syslog_facility",
+		"syslog_ident",
+		"tcp_keepalives_count",
+		"tcp_keepalives_idle",
+		"tcp_keepalives_interval",
+		"temp_file_limit",
+		"temp_tablespaces",
```
[the remark this answers is not in the document]
Comment (weinan003) [reply src/include/utils/unsync_guc_name.h]: Yes, I totally agree with you. I will go through all guc and evaluate whether should move to sync in next PR. I think it is easy to review. At present, these two arrays are directly core dump via guc name.

[the remark this answers is not in the document]
Comment (weinan003) [reply src/include/utils/unsync_guc_name.h]: Oh, sorry for miss understanding. I just realize what's happening. Thank you for notice. I will change it.

**Comment by ashwinstar** [comment]
Please can you add description on what would be the flow for developer for
- adding a GUC
- removing a GUC
- renaming a GUC
after this PR is committed, as it's not clear. Things like will I get compiler error if adding new guc but miss to add it to predefined array?
Also, do we know what guidelines need to followed to decide which array to add the guc to. That would help to get more weigh in and input from folks.

**Comment by weinan003** [comment]
> Please can you add description on what would be the flow for developer for
> 
> * adding a GUC
> * removing a GUC
> * renaming a GUC
>   after this PR is committed, as it's not clear. Things like will I get compiler error if adding new guc but miss to add it to predefined array?
>   Also, do we know what guidelines need to followed to decide which array to add the guc to. That would help to get more weigh in and input from folks.

Yes, I will add these in README

**Pull request merged**